Renesas HS300x Sensor Support

iandmorris opened the pull request:
Added driver and sample application for the Renesas HS3001 and HS3003 temperature/humidity sensors.

Tested using a Nucleo-F401RE, Click shield for Nucleo-64 and Temp&Hum Click 18 (HS3003) and Temp& Hum Click 17 (HS3001). The only difference between the HS3001 and HS3003 is the accuracy of the relative humidity measurements.

iandmorris committed:
drivers: sensor: hs300x: Add driver for Renesas HS300x sensors

Adds support for Renesas HS3001 and HS3003 temperature/humidity sensors
connected via an I2C bus.

Signed-off-by: Ian Morris <ian.d.morris@outlook.com>

iandmorris committed:
samples: sensor: dht_polling: Add generic dht sample application

This simple application periodically prints the temperature and humidity
measured by one or more digital humidity/temperature sensors.

Signed-off-by: Ian Morris <ian.d.morris@outlook.com>

kartben commented on drivers/sensor/hs300x/hs300x.c:
```diff
@@ -0,0 +1,159 @@
+		return -ENOTSUP;
+	}
+
+	rc = hs300x_start_measurement(dev);
```
Not sure it's worth a dedicated function?

iandmorris replied:
done

kartben commented on samples/sensor/hs300x/README.rst:
```diff
@@ -0,0 +1,65 @@
+.. _renesas_hs300x_sample:
+
+RENESAS_HS300X Sample
+#####################
```
Nit: typically only sensor model is used for naming sensor samples 

```suggestion
HS300X Sample
#############
```

iandmorris replied:
done

iandmorris commented:
I messed up when rebasing this branch with the main branch (very new to the Zephyr Git workflow) and can't seem to recover. I'm going to close this pull request, go away and learn/practice the work flow and then create a new PR that also addresses the comments made. The time you spent reviewing was not a waste, I'll be sure to incorporate the comments.

kartben commented:
> I messed up when rebasing this branch with the main branch (very new to the Zephyr Git workflow) and can't seem to recover. I'm going to close this pull request, go away and learn/practice the work flow and then create a new PR that also addresses the comments made. The time you spent reviewing was not a waste, I'll be sure to incorporate the comments.

git rebase origin/main --interactive and then just  delete that commit that's not yours?

iandmorris commented:
> > I messed up when rebasing this branch with the main branch (very new to the Zephyr Git workflow) and can't seem to recover. I'm going to close this pull request, go away and learn/practice the work flow and then create a new PR that also addresses the comments made. The time you spent reviewing was not a waste, I'll be sure to incorporate the comments.
> 
> git rebase origin/main --interactive and then just delete that commit that's not yours?

Thanks for the suggestion, however I have compounded the problem while trying to fix it (new to using git rebase). Quicker if i just create a new branch and PR (once I have added the features requested in the comments).

kartben commented:
I see. FWIW you should also be able to do a hard reset to any of your previous commits, which SHAs still show up on this PR or which you could also find via `git reflog`. Just thought it might be worthwhile for you to explore a bit more as there is almost always a solution that doesn't involve reopening a PR :)

iandmorris commented:
> I see. FWIW you should also be able to do a hard reset to any of your previous commits, which SHAs still show up on this PR or which you could also find via `git reflog`. Just thought it might be worthwhile for you to explore a bit more as there is almost always a solution that doesn't involve reopening a PR :)

Thanks for the tip about `git reflog` (and for your encouragement). I think I have resolved the issues I created...

iandmorris commented on drivers/sensor/hs300x/hs300x.c:
```diff
@@ -0,0 +1,159 @@
+
+	if (!device_is_ready(cfg->bus.bus)) {
+		LOG_ERR("I2C dev %s not ready", cfg->bus.bus->name);
+		return -EINVAL;
```
done

iandmorris commented on drivers/sensor/hs300x/hs300x.c:
```diff
@@ -0,0 +1,159 @@
+
+	/*
+	 * According to datasheet maximum time to make temperature and humidity
+	 * measurements is 33ms, add a little safety margin...
```
Making the resolution configurable is something I want to do. However, on HS300x sensors this is not as straightforward as simply writing to a register. The resolution is stored in non-volatile memory and to change it you must put the device into programming mode by cycling the power and then sending a specific command (within a certain time). 

Unfortunately, I don’t have time to figure these things out at present.  I was hoping to contribute this basic driver first (which I think is still useful), and then come back and add the above functionality when I have more time (after my project is complete). 

Hope this is an acceptable approach…?

iandmorris replied:
@kartben is this ok with you?

kartben replied:
sure!

MaureenHelm commented on drivers/sensor/hs300x/hs300x.c:
```diff
@@ -0,0 +1,153 @@
+{
+	const struct hs300x_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->bus.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(cfg->bus)) {
```

iandmorris replied:
done

MaureenHelm commented on drivers/sensor/hs300x/hs300x.h:
```diff
@@ -0,0 +1,21 @@
+
+#include <zephyr/device.h>
+
+struct hs300x_config {
+	struct i2c_dt_spec bus;
+};
+
+struct hs300x_data {
+	int16_t t_sample;
+	uint16_t rh_sample;
+};
```
Not enough here to justify having a separate private include file. Please move into `hs300x.c`

iandmorris replied:
done

MaureenHelm commented on samples/sensor/hs300x/README.rst:
```diff
@@ -0,0 +1,65 @@
+.. _hs300x_sample:
+
+HS300X Sample
```
We want to stop introducing new driver-specific sensor samples, and instead use more generic ones for each sensor type. See `samples/sensor/accel_polling` and `samples/sensor/magn_polling` for example

iandmorris replied:
makes sense. do you want me to create a generic sample for temperature and humidity sensors, based on one of these samples? something like ```samples/sensor/temp_hum_polling``` perhaps?

MaureenHelm replied:
That would be great, thank you

kartben replied:
Maybe simply repurpose the `dht` sample?

iandmorris replied:
ok, I will look at creating a generic temperature/humidity sample based on the existing ```dht```sample.

iandmorris commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -771,3 +771,8 @@ test_i2c_hm330x@73 {
+test_i2c_hs300x: hs300x@44 {
+	compatible = "renesas,hs300x";
+	reg = <0x44>;
+};
```
thanks, hadn't realized the address should be incremented.

MaureenHelm reviewed (commented):
Needs a rebase

MaureenHelm commented:
@kartben please revisit

kartben commented on samples/sensor/dht_polling/src/main.c:
```diff
@@ -0,0 +1,61 @@
+				break;
+			}
+
+			printk("%16s: temp is %d.%d oC humidity is %d.%d %%RH\n", dev->name,
```
```suggestion
			printk("%16s: temp is %d.%d °C humidity is %d.%d %%RH\n", dev->name,
```

Also, maybe the format should be `%d.%02d`, having so many digits feels odd.

iandmorris replied:
I changed ```oC``` to ```°C```, however when I tried the ```%d.%02d``` modification I found it did not do what I thought it would... After some reading it seems that (unless I have missed something) the width format specifier defines the minimum number of characters to be printed but it does not truncate the value if it is larger. Therefore there was no change in the output (in most cases) and so I left formatting as it was...

kartben replied:
Ah! of course... my bad. So also need to divide val2 by 10000, right?

https://github.com/zephyrproject-rtos/zephyr/blob/191e47bc8633a41ffd88ba64d740ea085bd68c0e/samples/shields/npm1300_ek/src/main.c#L96

kartben commented on samples/sensor/dht_polling/README.rst:
```diff
@@ -0,0 +1,49 @@
+
+.. code-block:: console
+
+       hs300x@44: temp is 25.31129 oC humidity is 30.391259 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.446194 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.372947 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.391259 %RH
+       hs300x@44: temp is 25.31129 oC humidity is 30.372947 %RH
+       hs300x@44: temp is 25.31129 oC humidity is 30.354635 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.372947 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.372947 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.391259 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.446194 %RH
+       hs300x@44: temp is 25.51272 oC humidity is 30.531648 %RH
```
update to use "°" and possibly the newer format as per my other comment.

iandmorris replied:
done, but see comment above.

kartben commented on samples/sensor/dht_polling/sample.yaml:
```diff
@@ -0,0 +1,20 @@
+    harness_config:
+      type: one_line
+      regex:
+        - "[0-9A-Za-z_,+-.]*@[0-9A-Fa-f]*: temp is (.*) oC humidity is (.*) %RH"
```
ditto

iandmorris replied:
done

kartben reviewed (changes requested):
Sorry for the delay! The new generic sample is great, thanks for this! 
A few remaining comments that are hopefully not too nitpicky :)

kartben commented on samples/sensor/dht_polling/src/main.c:
```diff
@@ -0,0 +1,62 @@
+				break;
+			}
+
+			printk("%16s: temp is %d.%02d °C humidity is %d.%02d %%RH\n",
+							dev->name, temp.val1, temp.val2 / 10000,
+							hum.val1, hum.val2 / 10000);
```
Just realized that many other samples actually use `sensor_value_to_double(&val))` and then a float formatter like `%.2f`. It does require `CONFIG_CBPRINTF_FP_SUPPORT=y` as well. But I'm fine with having this sample show an alternate way of doing things.

kartben reviewed (dismissed):
Thanks! My latest comment is not asking you for yet another update, to be clear :)

@MaureenHelm we should probably now just remove the existing `/dht` sample, right?

MaureenHelm reviewed (dismissed):
> we should probably now just remove the existing /dht sample, right?

Yes, but it would be nice to port `boards/nrf52dk_nrf52832.overlay` in the process

carlescufi commented:
@iandmorris please rebase to solve the conflict.
